즉시 교환 - 물품 제안 생성

tesbro1194 opened the pull request:
## 개요

<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
즉시 교환 - 물품 제안 생성

 - List 의 크기 확인하는 로직 없음 -> 제안 물품 1~3개인데 프론트에서 가능할 것으로 생각했습니다
 - 교환의 상태가 PENDING 인 교환에만 제안 가능
 - 제안 물품의 상태가 PENDING 인 제안 물품으로만 제안 가능
 - 그외 제가 개발하기 편하게 남겨놓은 주석들 있습니다. 해당 주석에 대한 개발이 완료되면 지우겠습니다
 
<!---- Resolves: #47 -->

## PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

tesbro1194 committed:
refactor: 디렉토리 변경

[#47]

tesbro1194 committed:
feat: 즉시 교환 - 제안 생성

[#47]

tesbro1194 committed:
refactor: 상태값 확인 코드 위임

[#47]

kimwanyoung commented on src/main/java/com/barter/domain/trade/immediatetrade/dto/request/CreateTradeSuggestProductReqDto.java:
```diff
@@ -0,0 +1,15 @@
+@Getter
+public class CreateTradeSuggestProductReqDto {
+	@NotBlank
+	List<SuggestedProduct> suggestedProductList = new ArrayList<>();
```
객체로 넘어오기보다는 Id로 넘어올 것 같다는 생각이 들어요! 객체로 넘기려면 프론트에서 Product에서 모든 정보를 갖고있어야하는데, 요청하는 화면에서는 간략한 정보만 존재할 거 같습니다.

그리고 변수 네이밍에서 자료형이 이미 List이기 때문에 중복 표현은 지양하는 것이 관례로 알고있어요!
위 경우에는 `suggestedProducts`정도로 네이밍을 하면 좋을 것 같아요!

tesbro1194 replied:
완영님 의견 반영

kimwanyoung commented on src/main/java/com/barter/domain/trade/immediatetrade/service/ImmediateTradeService.java:
```diff
@@ -84,4 +91,34 @@ public String delete(Long tradeId) {
+			throw new IllegalStateException("해당 교환에 제안할 수 없습니다.");
+		}
+
+		for (SuggestedProduct product : reqDto.getSuggestedProductList()) {
+			if (!product.validateProductStatus(product.getStatus())) { // PENDING 상태인 물품으로만 제안 가능
+				throw new IllegalArgumentException("해당 상품으로 제안하실 수 없습니다.");
+			}
+
+			TradeProduct tradeProduct = TradeProduct.builder()
+				.suggestedProduct(product)
+				.tradeType(TradeType.IMMEDIATE)
+				.build();
+
+			tradeProductRepository.save(tradeProduct);
+		}
```
여러건을 저장할 때에는 save() 보다 saveAll()을 사용하시는 것을 추천드립니다! 성능적으로 saveAll()이 유리하다고 알고있습니다

https://maivve.tistory.com/342

tesbro1194 replied:
완영님 의견 반영

kimwanyoung reviewed (approved):
고생하셨어요!

wootaepark reviewed (approved):
수고하셨습니다.

tesbro1194 committed:
refactor: comment 반영

- `SuggestedProduct -> Long`
- `suggestedProductList -> suggestedProductIds`
- `save -> saveAll`

[#47]

tesbro1194 committed:
Merge branch 'develop' into feat/immediate-trade/47
